Storage/transaction_policy.h

```diff
@@ -146,19 +149,35 @@ struct Synchronous final {
       return future;                                                                         // LCOV_EXCL_LINE
     }
     bool successful = false;
+    T_RESULT f1_result;
     try {
-      f2(f1());
+      f1_result = f1();
+      f2(std::move(f1_result));
       successful = true;
-    } catch (std::exception&) {  // LCOV_EXCL_LINE
-      journal_.Rollback();       // LCOV_EXCL_LINE
+    } catch (StorageRollbackExceptionWithValue<T_RESULT> e) {
+      // Transaction was rolled back, but returned a value, which we try to pass again to `f2`.
+      journal_.Rollback();
+      f2(std::move(e.value));
+      promise.set_value(TransactionResult<void>::RolledBack(OptionalResultMissing()));
+    } catch (StorageRollbackExceptionWithNoValue) {
+      // Transaction was rolled back and returned nothing we can pass to `f2`.
+      journal_.Rollback();
+      promise.set_value(TransactionResult<void>::RolledBack(OptionalResultMissing()));
+    } catch (...) {  // The exception is captured with `std::current_exception()` below.
+      // LCOV_EXCL_START
+      journal_.Rollback();
+      try {
+        promise.set_exception(std::current_exception());
+      } catch (const std::exception& e) {
+        std::cerr << "Storage internal error in Synchronous::Transaction: " << e.what() << std::endl;
```

Comment:
Some `Error during 'promise.set_exception()`` message would be more descriptive here.